lib/internal/http2/core.js

```diff
@@ -868,6 +870,9 @@ class Http2Session extends EventEmitter {
 
     if (!socket._handle || !socket._handle._externalStream) {
       socket = new StreamWrap(socket);
+      socket.on('error', (err) => {
```

Comment:
I'm not sure what the proper error handling would look like here. I think it might need to do more than `emit` and it should also be aware of whether the `socketOnError` already fired previously.

Maybe something like this would be more accurate:

```js
const wrappedSocket = socket;
socket = new StreamWrap(wrappedSocket);
socket.on('error', socketOnError.bind(wrappedSocket));
```

This would avoid the JSStreamWrap propagating an already handled error from the underlying socket and emitting on the session twice.

/cc @jasnell @mcollina @addaleax

Reply:
I also think this might need to handle `socket.on('close')` in a similar manner. So something like:

```js
socket.on('close', socketOnClose.bind(wrappedSocket));
```

Reply:
Proper error handling here is a simple .on('error', () => {}) as the underlying stream is already handled before hitting this switch

Reply:
/cc @ryzokuken

Reply:
> Proper error handling here is a simple .on('error', () => {}) as the underlying stream is already handled before hitting this switch

That presumes that the wrap can't error itself...

Reply:
Read the source, https://github.com/nodejs/node/blob/master/lib/internal/wrap_js_stream.js. All errors originate from the underlying stream unless I'm mistaken

Reply:
@mafintosh see line 56 for example but also the fact that it inherits the Socket means there might be more errors possible.

Reply:
@mafintosh @apapirovski I'm sorry that I'm not in a position to make a decent guess here. Could you help me reach a conclusion regarding what exactly needs to be done here?

Reply:
I would like input from @mcollina and @jasnell if possible. I think the following is correct:

```js
const wrappedSocket = socket;
socket = new StreamWrap(wrappedSocket);
socket.on('error', socketOnError.bind(wrappedSocket));
socket.on('close', socketOnClose.bind(wrappedSocket));

Reply:
that will trigger the error twice, as the wrappedSocket already triggers `socketOnClose` and `socketOnError`.

I think it would be fine to trigger an error if the wrappedSocket is in objectMode immediately as that seems to be the only case where the streamwrap emits a non-forwarded error? And then `socket.on('error', () => {})` the rest, as they are already handled

Reply:
Looking at the http2 code it seems ok to call socketOnError more than once as that simply calls destroy which should guard against that. The onClose method has a bit more going on so can tell if that is safe to do there, but i don't think we need to forward that in either case?

so

``` js
const wrappedSocket = socket;
socket = new StreamWrap(wrappedSocket);
socket.on('error', socketOnError.bind(wrappedSocket));
```

?

Reply:
> that will trigger the error twice, as the wrappedSocket already triggers socketOnClose and socketOnError.

It will just bail because `socketOnError` removes the session assignment in a sync call. I'm also not sure that the error in `objectMode` is the only one. The wrap is a socket in itself, it can throw other related errors.

Reply:
And I think you're probably right re: leaving out `close` listener. So the thing posted above seems good to me.